interfaces/prompting/requestrules/requestrules.go

```diff
@@ -0,0 +1,997 @@
+// records a notice for each one.
+//
+// The caller must ensure that the database lock is held for writing.
+func (rdb *RuleDB) removeRulesInternal(user uint32, rules []*Rule) error {
+	for _, rule := range rules {
+		rdb.removeExistingRuleNoError(rule)
+		data := map[string]string{"removed": "removed"}
+		rdb.notifyRule(user, rule.ID, data)
+	}
+	return rdb.save()
+}
+
+// RemoveRulesForSnapInterface removes all rules pertaining to the given snap
+// and interface for the user with the given user ID.
+func (rdb *RuleDB) RemoveRulesForSnapInterface(user uint32, snap string, iface string) ([]*Rule, error) {
+	rdb.mutex.Lock()
+	defer rdb.mutex.Unlock()
+	ruleFilter := func(rule *Rule) bool {
+		return rule.User == user && rule.Snap == snap && rule.Interface == iface
+	}
+	rules := rdb.rulesInternal(ruleFilter)
+	if err := rdb.removeRulesInternal(user, rules); err != nil {
+		return nil, err
+	}
+	return rules, nil
+}
+
+// PatchRule modifies the rule with the given ID by updating the rule's fields
+// corresponding to any of the given parameters which are set/non-empty.
+//
+// Any of the parameters which are equal to the default/unset value for their
+// types are left unchanged from the existing rule. Even if the given new rule
+// contents exactly match the existing rule contents, the timestamp of the rule
+// is updated to the current time. If there is any error while modifying the
+// rule, the rule is rolled back to its previous unmodified state, leaving the
+// database unchanged. If the database is changed, it is saved to disk.
+func (rdb *RuleDB) PatchRule(user uint32, id prompting.IDType, constraints *prompting.Constraints, outcome prompting.OutcomeType, lifespan prompting.LifespanType, duration string) (r *Rule, err error) {
+	rdb.mutex.Lock()
+	defer rdb.mutex.Unlock()
+	needToSave := false
+	changeOccurred := false
+
+	defer func() {
+		switch {
+		case changeOccurred:
+			rdb.notifyRule(user, id, nil)
+			fallthrough
+		case needToSave:
+			saveErr := rdb.save()
+			err = errorsJoin(err, saveErr)
+		}
+	}()
+
+	origRule, err := rdb.ruleWithIDForUser(user, id)
+	if err != nil {
+		return nil, err
+	}
+	if constraints == nil {
+		constraints = origRule.Constraints
+	}
+	if outcome == prompting.OutcomeUnset {
+		outcome = origRule.Outcome
+	}
+	if lifespan == prompting.LifespanUnset {
+		lifespan = origRule.Lifespan
+	}
+	if err = rdb.removeRuleFromTree(origRule); err != nil {
+		// If error occurs, rule is fully removed from tree
+		changeOccurred = true
+		return nil, err
+	}
+	newRule, err := rdb.populateNewRule(user, origRule.Snap, origRule.Interface, constraints, outcome, lifespan, duration)
+	if err != nil {
+		rdb.addRuleToTree(origRule) // ignore any new error, should not occur
+		// origRule was successfully removed before, so it should now be able
+		// to be successfully re-added without error, and all is unchanged.
+		changeOccurred = false
+		return nil, err
+	}
```

Comment:
Want to populate rule before touching the existing rule, in case validation fails.